adcc/ExcitedStates.py

```diff
@@ -657,6 +658,60 @@ def describe_amplitudes(self, tolerance=0.01, index_format=None):
                 ret += separator
         return ret[:-1]
 
+    def to_dataframe(self):
+        """
+        Exports the ExcitedStates object as :class:`pandas.DataFrame`.
+        Atomic units are used for all values.
+        """
+        propkeys = self.excitation_property_keys
+        propkeys.extend(self.excitation_energy_corrections.keys())
+        data = {
+            "excitation": np.arange(0, self.size, dtype=int),
+            "kind": np.tile(self.kind, self.size)
+        }
+        for key in propkeys:
+            try:
+                d = getattr(self, key)
+            except NotImplementedError:
+                # some properties are not available for every backend
+                continue
+            if not isinstance(d, np.ndarray):
+                continue
+            if not np.issubdtype(d.dtype, np.number):
+                continue
+            if d.ndim == 1:
+                data[key] = d
+            elif d.ndim == 2 and d.shape[1] == 3:
+                for i, p in enumerate(["x", "y", "z"]):
+                    data[f"{key}_{p}"] = d[:, i]
+            elif d.ndim > 2:
+                raise TypeError(f"Exporting NumPy array with shape {d.shape}"
+                                " not supported.")
```

Comment:
It just occurred to me that this should maybe be a warning followed by a continue, since the same reasoning of forward-compatibility applies. Also I'd add the `key` to the error message.

Reply:
Yeah that's nice!